daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala

```diff
@@ -92,21 +92,42 @@ sealed trait DINode {
     else asComplex
 
   def isDefaulted: Boolean
+  def isHidden: Boolean
 
   def children: Stream[DINode]
   def totalElementCount: Long
   def namedQName: NamedQName
   def erd: ElementRuntimeData
 
+  def infosetWalkerBlockCount: Int
+
   /**
    * Can treat any DINode, even simple ones, as a container of other nodes.
    * This simplifies walking an infoset.
    */
   def contents: IndexedSeq[DINode]
   final def numChildren = contents.length
 
-  def visit(handler: InfosetOutputter, removeHidden: Boolean = true): Unit
+  /**
+   * Returns the DINode that contains this DINode. Note that this behavior is
+   * slightly different from the "parent". The parent method always returns a
+   * DIComplex, even if this DINode is in an array. This usually makes sense
+   * and is what the caller intended when they call parent.
+   *
+   * However, we sometimes want the exact DINode that contains this DINode. So
+   * if this DINode is a DIElement in an array, then this returns the DIArray.
+   * Otherwise (i.e. DIArray or non-array DIElement) then it should return the
+   * DIComplex parent.
```

Comment:
Since the root is not an array (It's a DIDocument) it just returns parent, which I think is just null for DIDocument/DIComplex. I'll make this more clear.